sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala

```diff
@@ -184,4 +184,17 @@ abstract class ExternalCatalog {
 
   def listFunctions(db: String, pattern: String): Seq[String]
 
+  // --------------------------------------------------------------------------
+  // Resources
+  // --------------------------------------------------------------------------
+
+  /**
+   * Add a JAR resource to the underlying external catalog for DDL (e.g. CREATE TABLE) and DML
+   * (e.g., LOAD TABLE) operations.
+   *
+   * For example, when users create a Hive serde table, they can specify a custom
+   * Serializer-Deserializer (SerDe) class. When Hive metastore is unable to access the custom SerDe
+   * JAR (e.g., not on the Hive classpath), the JAR file must be added at runtime using this API.
+   */
+  def addJar(path: String): Unit
```

Comment:
Let me copy and paste the reasons why users need the customization capabilities:
- When to add a new File Format?

> User has files with special file formats not supported by Hive yet, and users don’t want to convert the files before loading into Hive.
> User has a more efficient way of storing data on disk.
- When to add a new File Format?

- When to add a new SerDe?

> User has data with special serialized format not supported by Hive yet, and users don’t want to convert the data before loading into Hive.
> User has a more efficient way of serializing the data on disk.

Also, the existing UDF, UDAF and UDTF can be reused without any rewriting.